### PR TITLE
#235 Add BatchValidator for category sync

### PR DESCRIPTION
**Pull request by salander85**
#### Summary
Implement a CategoryBatchValidator to improve caching in category sync 

#### Relevant Issues
 https://github.com/commercetools/commercetools-sync-java/issues/235

#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [ ] Documentation
- [ ] Add Release Notes entry.

#### Hints for Review
**NOTE**: CodeCov is failing because exception handling is never reached. As we need to come up with a new solution in the ReferenceResolvers this is ignored for now. Fixing this issue should be part of the second part of the ticket which is generalization of the Validator.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/categories/helpers/CategoryBatchValidator.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                    validKeys.add(categoryDraft.getKey());
+                    validKeys.add(parent.getKey());
+                }
+            }
+        }
+
+        return ImmutablePair.of(validDrafts, validKeys);
+    }
+
+    private boolean validateCategoryDraft(final CategoryDraft categoryDraft) {
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/categories/helpers/CategoryBatchValidator.java:76]:
```suggestion
    private boolean isValidCategoryDraft(final CategoryDraft categoryDraft) {
```

File: src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java [hunk 2]
```diff
@@ -36,38 +34,25 @@
...
-            .thenAccept(result -> isCached = true)
-            .thenApply(result -> keyToIdCache);
+    public CompletionStage<Map<String, String>> cacheKeysToIds(@Nonnull final Set<String> categoryKeys) {
+        return cacheKeysToIds(
+                categoryKeys, keysNotCached -> CategoryQuery
+                    .of()
+                    .withPredicates(buildCategoryKeysQueryPredicate(keysNotCached)));
+    }
+
+    QueryPredicate<Category> buildCategoryKeysQueryPredicate(@Nonnull final Set<String> categoryKeys) {
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java:50]: would it be better to make it private ?

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 3]
```diff
@@ -47,7 +47,6 @@
     Q extends MetaModelQueryDsl<U, Q, M, E>, M, E> {
 
     final S syncOptions;
-    boolean isCached = false;
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java:50]: 🎉

File: src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java (hunk 2 above)
Comment (salander85) [reply src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java:50]: ah, good hint. wanted to write a test for it but forgot :) 
actually there are no UnitTests for this class, should i add some?

File: src/main/java/com/commercetools/sync/categories/helpers/CategoryBatchValidator.java (hunk 1 above)
Comment (salander85) [reply src/main/java/com/commercetools/sync/categories/helpers/CategoryBatchValidator.java]: 7b7b3f7ece78f82b9779adbfa78c66f4c2e6a8bd

File: src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java (hunk 2 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/services/impl/CategoryServiceImpl.java:50]: Not for now, probably it's covered by integration tests.

**Review by ahmetoz — APPROVED** [review]

**Comment by ahmetoz** [comment]
Will be covered by #574

**Pull request closed without merging**